src/v/cluster/cloud_metadata/cluster_recovery_backend.cc

```diff
@@ -0,0 +1,427 @@
+                } catch (...) {
+                    auto eptr = std::current_exception();
+                    if (ssx::is_shutdown_exception(eptr)) {
+                        vlog(
+                          clusterlog.debug,
+                          "Shutdown error caught while recovering: {}",
+                          eptr);
+                    } else {
+                        vlog(
+                          clusterlog.error,
+                          "Unexpected error caught while recovering: {}",
+                          eptr);
+                    }
+                }
+            }
+        }
+    }
+}
+
+ss::future<> cluster_recovery_backend::recover_until_term_change() {
+    if (!_raft0->is_leader()) {
+        co_return;
+    }
+    ss::abort_source term_as;
+    _term_as = term_as;
+    auto reset_term_as = ss::defer([this] { _term_as.reset(); });
+    auto synced_term_opt = co_await _recovery_manager.sync_leader(term_as);
+    if (!synced_term_opt.has_value()) {
+        co_return;
+    }
+    auto synced_term = synced_term_opt.value();
+    if (!_recovery_table.local().is_recovery_active()) {
+        co_return;
+    }
+    auto recovery_state
+      = _recovery_table.local().current_recovery().value().get();
+    if (may_require_controller_recovery(recovery_state.stage)) {
+        auto controller_snap = co_await find_controller_snapshot_in_bucket(
+          recovery_state.bucket);
+        if (!controller_snap.has_value()) {
+            vlog(
+              clusterlog.error,
+              "Failed to download controller snapshot from bucket: {}",
+              recovery_state.bucket);
+            co_return;
+        }
+        vlog(
+          clusterlog.info,
+          "Downloaded controller snapshot. Proceeding with reconciliation...");
+
+        // We may need to restore state from the controller snapshot.
+        cloud_metadata::controller_snapshot_reconciler reconciler(
+          _recovery_table.local(),
+          config::shard_local_cfg(),
+          _features,
+          _creds,
+          _acls,
+          _topics);
+        auto controller_actions = reconciler.get_actions(
+          controller_snap.value());
+        vlog(
+          clusterlog.info,
+          "Controller recovery will proceed in {} stages",
+          controller_actions.stages.size());
+        auto err = co_await apply_controller_actions_in_term(
+          synced_term, std::move(controller_actions));
+        if (err != cluster::errc::success) {
+            co_return;
+        }
+    }
+    synced_term_opt = co_await _recovery_manager.sync_leader(term_as);
+    if (
+      !synced_term_opt.has_value() || synced_term_opt.value() != synced_term) {
+        co_return;
+    }
+    if (!_recovery_table.local().is_recovery_active()) {
+        co_return;
+    }
+    // All done! Record success.
+    auto errc = co_await _recovery_manager.replicate_update(
+      synced_term, recovery_stage::complete);
```

Comment:
i think we should pass the `synced_term` to `replicae_and_wait` as i mentioned in earlier comment.